lib/browser/source.js

```diff
@@ -11,7 +11,95 @@ class FileSource {
   close() {}
 }
 
-export function getSource(input) {
+class StreamSource {
+  constructor(reader, chunkSize) {
+    // Ensure that chunkSize is an integer and not something else or Infinity.
+    chunkSize = +chunkSize;
+    if (!isFinite(chunkSize)) {
+      throw new Error("cannot create source for stream without a finite value for the `chunkSize` option");
+    }
+    this._chunkSize = chunkSize;
+
+    this._buffer = undefined;
+    this._bufferOffset = 0;
+    this._reader = reader;
+    this._done = false;
+  }
+
+  slice(start, end, callback) {
+    if (start < this._bufferOffset) {
+      throw new Error("requested data is before the reader's current offset");
+    }
+
+    return this._readUntilEnoughDataOrDone(start, end, callback);
+  }
+
+  _readUntilEnoughDataOrDone(start, end, callback) {
+    const hasEnoughData = end <= this._bufferOffset + len(this._buffer);
+    if (this._done || hasEnoughData) {
+      callback(this._getDataFromBuffer(start, end));
+      return;
+    }
+    this._reader.read().then(({ value, done }) => {
```

Comment:
Could you please add error handling in the case that this promise gets rejected?